synapse/http/server.py

```diff
@@ -82,6 +83,52 @@
 </html>
 """
 
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+_cancellable_method_names = frozenset(
+    {
+        # `RestServlet`, `BaseFederationServlet` and `BaseFederationServerServlet`
+        # methods
+        "on_GET",
+        "on_PUT",
+        "on_POST",
+        "on_DELETE",
+        # `_AsyncResource`, `DirectServeHtmlResource` and `DirectServeJsonResource`
+        # methods
+        "_async_render_GET",
+        "_async_render_PUT",
+        "_async_render_POST",
+        "_async_render_DELETE",
+        "_async_render_OPTIONS",
+        # `ReplicationEndpoint` methods
+        "_handle_request",
+    }
+)
+
+
+def cancellable(method: F) -> F:
+    """Marks a servlet method as cancellable.
```

Comment:
it'd be nice to expand on this a little - ie, what does it mean to be "cancellable"? (Presumably: we can call `.cancel` on any of the `Deferred`s yielded by the coroutine to indicate the client is no longer interested in the response, and it will handle this cleanly)

Reply:
I've written some more words - please let me know what you think.